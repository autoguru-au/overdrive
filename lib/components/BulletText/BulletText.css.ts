import { style } from '@vanilla-extract/css';

import { overdriveTokens as vars } from '../../themes/theme.css';

export const bullet = style({
	width: vars.space['6'],
	height: vars.space['6'],
	lineHeight: vars.typography.size['2'].lineHeight,
});

export const primary = style({
	backgroundColor: vars.colours.intent.primary.background.mild,
});

export const secondary = style({
	backgroundColor: vars.colours.background.light,
});

export const primaryText = style({
	color: vars.colours.intent.primary.background.standard,
});

export const secondaryText = style({
	color: vars.typography.colour.dark,
});
