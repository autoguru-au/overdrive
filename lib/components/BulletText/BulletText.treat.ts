import { style } from 'treat';

const bulletSize = '24px';

export const root = style({
	listStyle: 'none',
});

export const customBullet = style({
	alignSelf: 'center',
	position: 'relative',
	flexShrink: 0,
});

export const bullet = style(theme => ({
	position: 'relative',
	flexShrink: 0,
	lineHeight: theme.typography.size['2'].lineHeight,
	width: bulletSize,
	height: bulletSize,
}));

export const bulletText = style(theme => ({
	fontSize: theme.typography.size['2'].fontSize,
	position: 'absolute',
	top: '50%',
	left: '50%',
	transform: 'translate(-50%, -50%)',
}));

export const variant = {
	primary: style(theme => ({
		color: theme.colours.gamut.green900,
		backgroundColor: theme.colours.gamut.green200,
	})),
	secondary: style(theme => ({
		color: theme.colours.gamut.gray900,
		backgroundColor: theme.colours.gamut.gray200,
	})),
};
