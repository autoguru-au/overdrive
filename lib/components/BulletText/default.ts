export { BulletText as default } from './BulletText';
