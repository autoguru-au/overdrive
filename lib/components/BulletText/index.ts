export { BulletText, EVariant as EBulletTextVariant } from './BulletText';
