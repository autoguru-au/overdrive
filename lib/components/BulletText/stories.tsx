import { CheckIcon } from '@autoguru/icons';
import { boolean, select, text } from '@storybook/addon-knobs';
import * as React from 'react';

import { Icon } from '../Icon';
import { BulletText } from '.';

const baseProps = () => ({
	variant: select('Variant', ['primary', 'secondary'], 'primary'),
	bullet: text('Bullet', '?'),
	ordered: boolean('ordered', false),
});

export default {
	title: 'Components|BulletText',
	component: BulletText,
};

export const standard = () => (
	<BulletText {...baseProps()}>Hello World</BulletText>
);

export const withCustomElement = () => (
	<BulletText
		bullet={
			<span
				style={{
					width: '20px',
					height: '20px',
					backgroundColor: ' #00dd95',
					color: 'white',
					borderRadius: '50%',
					flexDirection: 'row',
					display: 'flex',
					alignContent: 'center',
					justifyContent: 'center',
					alignItems: 'center',
				}}>
				<Icon size="small" icon={CheckIcon} />
			</span>
		}>
		Hello World
	</BulletText>
);
