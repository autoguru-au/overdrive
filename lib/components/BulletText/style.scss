@import '../../theme/tokens/facade';

:root {
	--bullet--size: 24px;
}

.root {
	display: flex;
	align-content: center;
	align-items: center;
	flex-direction: row;
	box-sizing: border-box;
	margin: 0;
	padding: 0;
	list-style: none;
}

.bullet {
	line-height: var(--bullet--size);
	flex: var(--bullet--size) 0 0;
	height: var(--bullet--size);
	border-radius: 50%;

	.variant-primary & {
		color: getColour(green, 900);
		background-color: getColour(green, 200);
	}

	.variant-secondary & {
		color: getColour(gray, 900);
		background-color: getColour(gray, 200);
	}
}

.custom-bullet,
.bullet {
	position: relative;
	flex-shrink: 0;
	margin: 2px var(--global--space--4) 0 0;
}

.custom-bullet {
	align-self: center;
}

.bullet-text {
	font-size: 12px;
	position: absolute;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -50%);
}
