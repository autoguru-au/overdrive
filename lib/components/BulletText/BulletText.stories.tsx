import { CheckIcon } from '@autoguru/icons';
import { Meta, StoryObj } from '@storybook/react-vite';
import * as React from 'react';

import { Icon } from '../Icon/Icon';

import { BulletText } from './BulletText';

const meta = {
	title: 'Content/Bullet Text',
	component: BulletText,
	args: {
		as: 'div',
		children: 'Hello World',
		variant: 'primary',
	},
	argTypes: {
		bullet: {
			table: {
				type: { summary: 'Any custom react element' },
				defaultValue: void 0,
			},
			description: 'Any custom react element',
			control: {
				type: 'select',
			},
		},
	},
} satisfies Meta<typeof BulletText>;

export default meta;
type Story = StoryObj<typeof BulletText>;

export const Primary: Story = {};

export const Secondary: Story = {
	args: {
		variant: 'secondary',
	},
};

export const WithCustomBullet: Story = {
	args: {
		variant: 'primary',
		bullet: (
			<span
				style={{
					width: '20px',
					height: '20px',
					backgroundColor: ' #00dd95',
					color: 'white',
					borderRadius: '50%',
					flexDirection: 'row',
					display: 'flex',
					alignContent: 'center',
					justifyContent: 'center',
					alignItems: 'center',
				}}
			>
				<Icon size="small" icon={CheckIcon} />
			</span>
		),
	},
};
