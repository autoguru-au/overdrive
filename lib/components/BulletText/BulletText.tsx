import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, isValidElement, ReactNode } from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../Box';
import { Column, Columns } from '../Columns';
import { Text } from '../Typography/Text';
import * as styleRefs from './BulletText.treat';

interface Props {
	bullet?: ReactNode;
	variant?: keyof typeof styleRefs.variant;
	className?: string;
}

export const BulletText: FunctionComponent<Props> = ({
	variant = 'primary',
	className = '',
	children,
	bullet: Bullet = 'â€¢',
}) => {
	const styles = useStyles(styleRefs);

	return (
		<Columns
			noWrap
			space="3"
			className={clsx(styles.root, className)}
			is="li">
			{isValidElement(Bullet) ? (
				<Column alignSelf="centre" className={styles.customBullet}>
					{Bullet}
				</Column>
			) : (
				<Column alignSelf="centre">
					<Box
						className={clsx(styles.bullet, styles.variant[variant])}
						borderRadius="pill">
						<span className={styles.bulletText}>{Bullet}</span>
					</Box>
				</Column>
			)}
			<Column>
				<Text is="span">{children}</Text>
			</Column>
		</Columns>
	);
};
