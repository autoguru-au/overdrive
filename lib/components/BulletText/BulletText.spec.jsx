import { CheckIcon } from '@autoguru/icons';
import { render } from '@testing-library/react';
import * as React from 'react';

import { Icon } from '../Icon';
import { BulletText } from './BulletText';

const CustomBullet = () => (
	<span>
		<Icon size={14} icon={CheckIcon} />
	</span>
);

describe('<BulletText />', () => {
	it('should not throw', () =>
		expect(() => render(<BulletText />)).not.toThrow());

	it.skip('should pass on className to dom element', () => {
		expect(
			render(<BulletText className="bullet-class" />).container
				.firstChild,
		).toHaveClass('bullet-class');
	});

	it('should match snapshot for default bullet text', () => {
		expect(
			render(<BulletText>Hello World!</BulletText>).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for primary bullet text with custom bullet character', () => {
		expect(
			render(
				<BulletText variant="primary" bullet="?">
					Hello World!
				</BulletText>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for secondary bullet text with custom bullet character', () => {
		expect(
			render(
				<BulletText variant="secondary" bullet="?">
					Hello World!
				</BulletText>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for ordered bullet text', () => {
		expect(
			render(
				<BulletText ordered variant="primary" bullet="?">
					Hello World!
				</BulletText>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should match snapshot for custom bullet text', () => {
		expect(
			render(
				<BulletText bullet={<CustomBullet />}>Hello World</BulletText>,
			).container.firstChild,
		).toMatchSnapshot();
	});

	it('should pass the children a Text element', () => {
		expect(
			render(<BulletText>Hello World!</BulletText>).container,
		).toHaveTextContent('Hello World!');
	});

	it('should use li element for default bullet text', () => {
		expect(
			render(
				<BulletText>Hello World!</BulletText>,
			).container.querySelector('li'),
		).toBeInTheDocument();
	});

	it('should use • as default bullet', () => {
		expect(
			render(
				<BulletText>Hello World!</BulletText>,
			).container.querySelector('span'),
		).toHaveTextContent('•');
	});

	it('should use provided bullet character bullet', () => {
		expect(
			render(
				<BulletText bullet="#">Hello World!</BulletText>,
			).container.querySelector('span'),
		).toHaveTextContent('#');
	});
});
