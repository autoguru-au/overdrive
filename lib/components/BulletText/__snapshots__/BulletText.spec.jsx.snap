// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulletText /> should match snapshot for custom bullet text 1`] = `
<li
  class="root variantPrimary"
>
  <div
    class="customBullet"
  >
    <span
      style="width:17px;height:17px;background-color:#00dd95;fill:white;border-radius:50%;flex-direction:row;display:flex;align-content:center;justify-content:center;align-items:center"
    >
      <i
        class="root"
        style="width:14px;height:14px"
      >
        <svg
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 149.333l-256 256L74.667 288l30.08-30.08L192 344.96l225.92-225.707 30.08 30.08z"
          />
        </svg>
      </i>
    </span>
  </div>
  <p
    class="root"
  >
    Hello World
  </p>
</li>
`;

exports[`<BulletText /> should match snapshot for default bullet text 1`] = `
<li
  className="root variantPrimary"
>
  <div
    className="bullet"
  >
    <span
      className="bulletText"
    >
      â€¢
    </span>
  </div>
  <Text>
    Hello World!
  </Text>
</li>
`;

exports[`<BulletText /> should match snapshot for ordered bullet text 1`] = `
<ol
  className="root variantPrimary"
>
  <div
    className="bullet"
  >
    <span
      className="bulletText"
    >
      ?
    </span>
  </div>
  <Text>
    Hello World!
  </Text>
</ol>
`;

exports[`<BulletText /> should match snapshot for primary bullet text with custom bullet character 1`] = `
<li
  className="root variantPrimary"
>
  <div
    className="bullet"
  >
    <span
      className="bulletText"
    >
      ?
    </span>
  </div>
  <Text>
    Hello World!
  </Text>
</li>
`;

exports[`<BulletText /> should match snapshot for secondary bullet text with custom bullet character 1`] = `
<li
  className="root variantSecondary"
>
  <div
    className="bullet"
  >
    <span
      className="bulletText"
    >
      ?
    </span>
  </div>
  <Text>
    Hello World!
  </Text>
</li>
`;
